PublicMapping/district-builder-2
# Add script to convert GeoJSON to TopoJSON

## Overview

All district-related operations in the application will make use of hierarchical TopoJSON files for fast topologically-aware merging of polygons. This script takes a block level GeoJSON file as input, and outputs a TopoJSON file with a hierarchy of relevant geolevels and aggregated demographic data. The geolevel hierarchy and demographic data selection is dynamic and is controlled by command-line flags, so it can support any desired configuration (including any number of nested geolevels). Also, a flag exists to control whether or not the GeoJSON should be streamed in, which allows working with very large files (getting around a JSON.parse memory limitation).

Since there wasn't an already-existing section of the repo to run scripts such as this, I created a new command-line interface application named `manage` that could serve this purpose. It uses `oclif` from Heroku, which is a well thought out framework for creating CLIs. Because there's only the single command at the moment, and it's only to be run by developers, I haven't put in the work to containerize it. I think that may be a good task to do in the future, since I can foresee this CLI being used in production to execute long-running tasks. Note that code in the CLI may also use code from the `shared` module easily, however it's not being made use of at the moment.

### Demo

Counties created from blocks for both Pennsylvania and Delaware:

![PA_DE](https://user-images.githubusercontent.com/6386/75685890-55dba180-5c69-11ea-9ded-df76cdc6121d.png)

Counties created from blocks for California:

![CA](https://user-images.githubusercontent.com/6386/75685889-55430b00-5c69-11ea-94ad-0e40889f9825.png)

### Notes

I found the types defined by Topojson to be especially finicky, and often subtly inconsistent among functions. So unfortunately, there was a need to break out some `any` types, as well as introduce some mutation.

## Testing Instructions

- `cd src/manage`
- Download some GeoJSON. The files I tested with are a new PA file that has blockgroups defined, and old DE and CA files that don't have blockgroups. I will list the three commands I used to generate TopoJSON for these files in-particular. But test with whichever one(s) call to you:
    - `./bin/run process-geojson ~/DE.geojson -l block,tract,county`
    - `NODE_OPTIONS="--max-old-space-size=8192" ./bin/run process-geojson ~/PA.geojson -b`
    - `NODE_OPTIONS="--max-old-space-size=14336" ./bin/run process-geojson ~/CA.geojson -b -l block,tract,county`
- Now you should have a valid TopoJSON file in `output.json`. A good way to test everything worked correctly is to extract the counties from the TopoJSON and open up the GeoJSON in a tool such as QGIS. To do so, run: 
    - `topo2geo county=output-county.json < output.json`
- Click around the counties and ensure they look realistic (both in shape and in aggregated demographic data properties)

Closes #27
